### ISSUE TITLE
Add common trait for all frontends

### ISSUE THREAD

**Issue by gbaranski** [helix-editor/helix#618]
that would unblock #39

**Comment by sudormrfbin** [comment]
I'm going to try to implement this, and the following is more or less the design I'm thinking of.

- Make [`Compositor`](https://github.com/helix-editor/helix/blob/d84f8b5fdef71da87ee108db07ba1167fc6a769b/helix-term/src/compositor.rs#L74) a trait, which can be implemented by each frontend. It takes items which implement the [`Component` trait](https://github.com/helix-editor/helix/blob/d84f8b5fdef71da87ee108db07ba1167fc6a769b/helix-term/src/compositor.rs#L117) and [renders them](https://github.com/helix-editor/helix/blob/d84f8b5fdef71da87ee108db07ba1167fc6a769b/helix-term/src/compositor.rs#L144-L157) as layers on layers.

- Move [structs implmenting] `Component`s like `FilePicker`, `EditorView`, `Popup`, etc to `helix-view`. Now we have everything except the rendering logic in the base `helix-view` crate and frontends can handle rendering. This introduces some difficulties in the actual concrete code, which is explained below.

The `Compositor` and `Component` traits will be both defined in `helix-view`. The common "widgets" (items implementing `Component`) will also be in `helix-view`. These widgets will need to have the `Component` trait defined separately by each frontend. Since the frontend is another crate (like `helix-term`) we cannot do this:

```rust
// frontend crate
use helix_view::Component;
use helix_view::ui::FilePicker

// cannot implement a foreign trait for a foreign type
impl Component for FilePicker { /* ... */ }
```

We will need to use the [newtype pattern](https://stackoverflow.com/a/25415289/7115678) and define wrapper structs for all widgets in frontends, which can then implement `Component` and provide additional functionality if needed.

Any other ideas/suggestions welcome :)

**Comment by kirawi** [comment]
Closing in favor of https://github.com/helix-editor/helix/issues/5555
